### PR TITLE
Relay 7.0.0 adjustments

### PR DESCRIPTION
**Pull request by sorenhoyer**
Adjustments mostly base on https://github.com/facebook/relay/commit/2606f32674f75513726207ceb198c906c276e90c#diff-f5a244f9b04c4b73796c0bd3e1c43854 to try and add support for Relay 7

Todo:

- [x] Fix last remaining failing test related to fixtures/type-generator/derived-field.graphql
- [x] Clean up and refactor a bit
- [ ] Add missing types

**Expect to be done with the remaining todos by late tonight (GMT+1)**

First time I dive into this codebase, so feel free to help out with comments or commits on this PR, if you see something is wrong or just have time to lend a hand.

### CODE REVIEW

File: src/@types/relay-compiler.d.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  export interface Schema {
+    expectStringType: () => ScalarTypeID;
+    getEnumValues: (type: EnumTypeID) => string[];
+    getFieldName: (field: FieldID) => string;
+    getFields: (type: TypeID) => FieldID[];
+    getFieldType: (field: FieldID) => TypeID;
+    getListItemType: (type: TypeID) => TypeID;
+    getNullableType: (type: TypeID) => TypeID;
+    getTypeString: (type: TypeID) => string;
+    isAbstractType: (type: TypeID) => boolean;
```
Comment (smmoosavi) [inline src/@types/relay-compiler.d.ts]: we can use [typescript guards][type-guards] here

e.g.
```ts
isAbstractType: (type: TypeID) => type is (UnionType | InterfaceType);
isEnum: (type: TypeID) => type is EnumType;
``` 

[type-guards]: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

File: tsconfig.json [hunk 2]
```diff
@@ -22,7 +22,7 @@
     "downlevelIteration": true /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */,
     // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */
     /* Strict Type-Checking Options */
-    "strict": true /* Enable all strict type-checking options. */,
+    "strict": false /* Enable all strict type-checking options. */,
```
Comment (smmoosavi) [inline tsconfig.json]: I think you can add the following lines to `src/@types/relay-compiler.d.ts` and revert this file.

```ts
declare module "relay-compiler/lib/transforms/RefetchableFragmentTransform";
declare module "relay-compiler/lib/transforms/MaskTransform";
declare module "relay-compiler/lib/transforms/MatchTransform";
declare module "relay-compiler/lib/transforms/RelayDirectiveTransform";
```

**Comment by renanmav** [comment]
Delete your `package-lock.json`

Use yarn instead

**Comment by renanmav** [comment]
Sent you a PR tweaking some things.

https://github.com/sorenhoyer/relay-compiler-language-typescript/pull/1

**Comment by renanmav** [comment]
Just a reminder: in order to CI pass, we need to update the `@types/relay-compiler`. 

Currently, the `generator` function of [`TypeGenerator`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/relay-compiler/lib/language/RelayLanguagePluginInterface.d.ts#L234) its `generate: (node: Root | Fragment, options: TypeGeneratorOptions) => string;`

We need it to be `generate: (schema: Schema, node: Root | Fragment, options: TypeGeneratorOptions) => string;`

**Comment by sorenhoyer** [comment]
> Just a reminder: in order to CI pass, we need to update the `@types/relay-compiler`.
> 
> Currently, the `generator` function of [`TypeGenerator`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/relay-compiler/lib/language/RelayLanguagePluginInterface.d.ts#L234) its `generate: (node: Root | Fragment, options: TypeGeneratorOptions) => string;`
> 
> We need it to be `generate: (schema: Schema, node: Root | Fragment, options: TypeGeneratorOptions) => string;`

There's a PR for it here. Once that is approved, I will update the @types/relay dependency and remove the temporary types.

File: tsconfig.json (hunk 2 above)
Comment (sorenhoyer) [reply tsconfig.json]: If only it was that simple. With TS in strict mode, you'd get:

> Invalid module name in augmentation. Module 'relay-compiler/lib/transforms/RefetchableFragmentTransform' resolves to an untyped module at 'c:/code/relay-compiler-language-typescript/node_modules/relay-compiler/lib/transforms/RefetchableFragmentTransform.js', which cannot be augmented.

I guess we'll just have to wait for https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39665 to be merged.

File: example/package.json [hunk 3]
```diff
@@ -31,27 +31,28 @@
...
+    "react-relay": "^7.0.0",
+    "relay-runtime": "^7.0.0",
     "todomvc-app-css": "^2.1.0",
     "todomvc-common": "^1.0.3",
     "webpack": "^3.5.5",
     "webpack-dev-server": "^2.7.1",
     "whatwg-fetch": "2.0.3"
   },
   "devDependencies": {
+    "@types/node": "^12.11.7",
```
Comment (alloy) [inline example/package.json]: Why is this needed?

Comment (renanmav) [reply example/package.json]: When not installed, `yarn start` on example folder gives the following error:
```bash
ERROR in /Users/renan/www/relay-compiler-language-typescript/example/node_modules/@types/node/index.d.ts(200,11):
TS2300: Duplicate identifier 'IteratorResult'.
ERROR in /Users/renan/www/relay-compiler-language-typescript/example/node_modules/typescript/lib/lib.es2015.iterable.d.ts(41,6):
TS2300: Duplicate identifier 'IteratorResult'.
```

Solved upgrading `@types/node` suggested [here](https://github.com/microsoft/TypeScript/issues/32333#issuecomment-522408244).

Comment (alloy) [reply example/package.json]: Ok, so somehow we were already pulling in this package anyways, just an older version. Cool thanks 👍

**Comment by renanmav** [comment]
`@types/relay-compiler` 7.0.0 was released

**Comment by sorenhoyer** [comment]
I will make the last commits within a few hours.

**Comment by sorenhoyer** [comment]
Still 4 more tests fails, compared to running tests in master branch, so little more work is needed. :/

**Comment by sorenhoyer** [comment]
Down to just 1 more test failed than tests in master. If I remove fixtures/type-generator/derived-field.graphql it seems to work, weird :/!

**Comment by alloy** [comment]
Nice, you got this! 🙌 (But let us know if you need help.)

**Comment by sorenhoyer** [comment]
alloy thanks, much appreciated.

It should work now, but I will need to do some cleanup and some refactoring, formatting and linting and I'll make a PR with some additional types.

**Comment by sorenhoyer** [comment]
And why do I get " 3 failed, 54 passed, 57 total" tests when I run `yarn test` locally on this branch, and the same "3 failed, 54 passed, 57 total" tests when I clone down the current master branch? Travis-ci only reports "1 failed, 56 passed, 57 total", so why is this different when running locally?

**Comment by sorenhoyer** [comment]
Hopefully the only thing remaining is to add the missing types. Didn't get to do that today, but can make a PR tomorrow.

**Comment by sorenhoyer** [comment]
Oooops :scream:

Accidentally deleteted the old fork (should have been the DefinitelyTyped one instead!!!) that was linked to this PR 👏, so unfortunately had to abandon this one, make a new fork, copy all the changes (lucky I still had it locally) and make a new PR. 

For reference, here's the new PR: https://github.com/relay-tools/relay-compiler-language-typescript/pull/155

Closing this.
